src/Workspaces/CSharp/Portable/Editing/CSharpImportAdder.cs

```diff
@@ -66,138 +66,102 @@ protected override SyntaxNode MakeSafeToAddNamespaces(SyntaxNode root, IEnumerab
-                IEnumerable<IMethodSymbol> extensionMethods,
+            public ConflictWalker(
                 SemanticModel model,
-                Workspace workspace,
+                ImmutableArray<INamespaceSymbol> namespaceSymbols,
+                HashSet<INamespaceSymbol> conflictNamespaces,
                 CancellationToken cancellationToken)
+                : base(SyntaxWalkerDepth.StructuredTrivia)
             {
                 _model = model;
-                _workspace = workspace;
                 _cancellationToken = cancellationToken;
-                _namespaceMembers = new HashSet<string>(namespaceMembers.Select(x => x.Name));
-                _extensionMethods = new HashSet<string>(extensionMethods.Select(x => x.Name));
-            }
-
-            public override bool VisitIntoStructuredTrivia => true;
+                _conflictNamespaces = conflictNamespaces;
 
-            [return: NotNullIfNotNull("node")]
-            public override SyntaxNode? Visit(SyntaxNode? node) => base.Visit(node);
-
-            public override SyntaxNode VisitIdentifierName(IdentifierNameSyntax node)
-            {
-                // We only care about xml doc comments
-                var leadingTrivia = CanHaveDocComments(node) ? VisitList(node.GetLeadingTrivia()) : node.GetLeadingTrivia();
-
-                if (_namespaceMembers.Contains(node.Identifier.Text))
+                foreach (var ns in namespaceSymbols)
                 {
-                    var expanded = Simplifier.Expand<SyntaxNode>(node, _model, _workspace, cancellationToken: _cancellationToken);
-                    return expanded.WithLeadingTrivia(leadingTrivia);
+                    foreach (var type in ns.GetTypeMembers())
+                    {
+                        _namespaceMembers.Add((type.Name, type.Arity), ns);
+
+                        if (type.MightContainExtensionMethods)
+                        {
+                            foreach (var member in type.GetMembers())
+                            {
+                                if (member is IMethodSymbol method && method.IsExtensionMethod)
+                                    _extensionMethods.Add(method.Name, ns);
+                            }
+                        }
+                    }
                 }
-
-                return node.WithLeadingTrivia(leadingTrivia);
             }
 
-            public override SyntaxNode VisitGenericName(GenericNameSyntax node)
+            private void CheckName(NameSyntax node)
             {
-                // We only care about xml doc comments
-                var leadingTrivia = CanHaveDocComments(node) ? VisitList(node.GetLeadingTrivia()) : node.GetLeadingTrivia();
+                // Check to see if we have an standalone identifier (or identifier on the left of a dot). If so, if that
+                // identifier binds to a type, then we don't want to bring in any imports that would bring in the same
```

Comment:
> 

What about this case, where it doesn't bind to a named type, but adding an `using E` will make it ambiguous:

```csharp
using System;
using static D;
class C
{
    public static void Main()
    {
        Console.WriteLine(P.Length);
    }
}

class D
{
    public static string P = "   ";
}

namespace E
{
    public class P
    {
        public static int Length => 43;
    }
}
```

Reply:
For now i'm going to not handle this case.  Specifically:
1. you have to have a using static.  
2. you have to be referring to a member (not type) in that using static which will have the same name as a type that will be imported

Seems uncommon enough to not warrant special handling *for now*.  we can see if people actually run into this.